### PR TITLE
#352: BallistaContext::collect() logging is too noisy

### LINKED ISSUE #352: BallistaContext::collect() logging is too noisy
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

When running benchmarks, I see pages of this output and I lose the output I care about.

```
[2021-05-16T14:01:49Z INFO  ballista::context] Job FobBz8Q is running...
[2021-05-16T14:01:49Z INFO  ballista::context] Job FobBz8Q is running...
[2021-05-16T14:01:49Z INFO  ballista::context] Job FobBz8Q is running...
[2021-05-16T14:01:49Z INFO  ballista::context] Job FobBz8Q is running...
[2021-05-16T14:01:49Z INFO  ballista::context] Job FobBz8Q is running...
[2021-05-16T14:01:49Z INFO  ballista::context] Job FobBz8Q is running...
[2021-05-16T14:01:49Z INFO  ballista::context] Job FobBz8Q is running...
[2021-05-16T14:01:50Z INFO  ballista::context] Job FobBz8Q is running...
[2021-05-16T14:01:50Z INFO  ballista::context] Job FobBz8Q is running...
[2021-05-16T14:01:50Z INFO  ballista::context] Job FobBz8Q is running...
```

**Describe the solution you'd like**
We should either change the log level of this message to TRACE or consider a different approach to keeping the user informed about job status.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by jgoday**
# Which issue does this PR close?
Fix #352.

 # Rationale for this change
See #352 (logging too noisy inside BallistaContext::collect() inner loop).

# What changes are included in this PR?
Change info messages for trace inside collect loop.

# Are there any user-facing changes?
N/A

### CODE REVIEW

**Review by andygrove — COMMENTED** [review]
Thanks @jgoday. With this change we will never see output saying that the query is running unless we have trace logging enabled. Would it be possible to show the message with info logging on the first time we see that the job is running?

**Comment by jgoday** [comment]
> Thanks @jgoday. With this change we will never see output saying that the query is running unless we have trace logging enabled. Would it be possible to show the message with info logging on the first time we see that the job is running?

Ok, Is ok if we use a flag (first time a job_id is logged) and define a macro info_or_trace_if like ?

```rust

macro_rules! info_or_trace_if {
    ($first:ident, $($arg:tt)+) => (
        if $first {
            info!($($arg)+);
        } else {
            trace!($($arg)+);
        }
        $first = false;
    )
}
```

**Comment by andygrove** [comment]
Perhaps we should just log the status if the status has changed since the last time round the loop?

**Comment by jgoday** [comment]
@andygrove I have changed the condition to call info/trace to check the job's prev status.
Should we trace all loop iterations or just info! when the status has changed (so we don't need the info_or_trace_if macro) ?

**Comment by andygrove** [comment]
> @andygrove I have changed the condition to call info/trace to check the job's prev status.
> Should we trace all loop iterations or just info! when the status has changed (so we don't need the info_or_trace_if macro) ?

Nice. I think we can just log with `info!` and remove the macro in this case.

**Comment by jgoday** [comment]
> > @andygrove I have changed the condition to call info/trace to check the job's prev status.
> > Should we trace all loop iterations or just info! when the status has changed (so we don't need the info_or_trace_if macro) ?
> 
> Nice. I think we can just log with `info!` and remove the macro in this case.

Ok, I have changed it that way.

File: ballista/rust/client/src/context.rs [hunk 1]
```diff
@@ -198,14 +200,21 @@ impl BallistaContext {
                 DataFusionError::Internal("Received empty status message".to_owned())
             })?;
             let wait_future = tokio::time::sleep(Duration::from_millis(100));
+            let has_status_change = prev_status.map(|x| x != status).unwrap_or(false);
```
Comment (andygrove) [inline ballista/rust/client/src/context.rs:203]:
```suggestion
            let has_status_change = prev_status.map(|x| x != status).unwrap_or(true);
```
I think the default should be `true` here. If there was no previous status then the status must have changed?

Comment (jgoday) [reply ballista/rust/client/src/context.rs]: You are right, sorry for the mistake :(

**Review by andygrove — APPROVED** [review]
LGTM. I haven't tested this, but plan on doing so tonight. I will merge once I have confirmed that this works as expected.

Thanks @jgoday!

**Comment by andygrove** [comment]
```
[2021-05-27T00:55:04Z INFO  ballista::context] Connecting to Ballista scheduler at http://ballista-scheduler:50050
[2021-05-27T00:55:04Z INFO  ballista::context] Job kO3v67p still queued...
[2021-05-27T00:55:04Z INFO  ballista::context] Job kO3v67p is running...
Query 12 iteration 0 took 5751.5 ms
+------------+-----------------+----------------+
| l_shipmode | high_line_count | low_line_count |
+------------+-----------------+----------------+
| MAIL       | 6202            | 9324           |
| SHIP       | 6200            | 9262           |
+------------+-----------------+----------------+
Query 12 avg time: 5751.51 ms
```

:heart:

**Pull request merged**